src/Qir/qir-utils.ps1

```diff
@@ -93,10 +95,60 @@ function Build-CMakeProject {
+        #   >>> lib/QIR/CMakeFiles/qir-rt-support-obj.dir/QubitManager.cpp.obj has value MD_DynamicRelease
+        #   >>> clang_rt.ubsan_standalone_cxx-x86_64.lib(ubsan_type_hash_win.cc.obj) has value MT_StaticRelease
+        #   clang++: error: linker command failed with exit code 1 (use -v to see invocation)
+        $sanitizeFlags += " -fsanitize=undefined -fsanitize=float-divide-by-zero -fsanitize=unsigned-integer-overflow -fsanitize=implicit-conversion -fsanitize=local-bounds -fsanitize=nullability"
+        # TODO: 
+        #     For Win consider extra build configuration linking all libs staticly, enable `-fsanitize=undefined`, run the staticly linked tests.
+
+        if (-not ($IsMacOS))
+        {
+            # Safe Stack instrumentation (https://clang.llvm.org/docs/SafeStack.html):
+            #   No support for Win, Mac.
+            #       clang: error: unsupported option '-fsanitize=safe-stack' for target 'x86_64-apple-darwin19.6.0'
+            #   Linking a DSO with SafeStack is not currently supported. But compilation, linking, and test runs all succeed.
+            $sanitizeFlags += " -fsanitize=safe-stack"
+        }
+
+        ## Memory Sanitizer (https://clang.llvm.org/docs/MemorySanitizer.html)
+        ## Win: Not supported.
+        ##      clang: error: unsupported option '-fsanitize=memory' for target 'x86_64-pc-windows-msvc'
+        ## WSL: Complains for use-of-uninitialized-value in `catch2/catch.hpp` during initialization of global vars 
+        ##      (if run both as `pwsh Runtime/test-qir-runtime.ps1` (or Tests/test-qir-tests.ps1) and as standalone). 
+        ##      An update of `catch2/catch.hpp` to 2.13.6 (search for "go to the v2.x branch" at https://github.com/catchorg/Catch2) didn't help.
+        ##      Suppressing of the errors in the updated `catch2/catch.hpp` and standard library headers eventually bumps into errors reported in `memcmp`,
+        ##      suppressing of which does not work (https://github.com/google/sanitizers/issues/1429#issuecomment-876799463).
+        ##      Looks like MSan will not work until the libstdc++ is recompiled to be instrumented (https://clang.llvm.org/docs/MemorySanitizer.html#handling-external-code).
+        ##      Instrumenting libstdc++ during CI builds seems impractical (https://stackoverflow.com/a/22301584/6362941).
+        #$sanitizeFlags += " -fsanitize=memory -fsanitize-memory-track-origins=2"
+
+        # Common for all sanitizers:
+        $sanitizeFlags += " -fsanitize-blacklist="      # https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html#suppressing-errors-in-recompiled-code-ignorelist
+                                                        # https://releases.llvm.org/11.0.1/tools/clang/docs/SanitizerSpecialCaseList.html
+        $sanitizeFlags += (Join-Path $Path .. UBSan.ignore)
+
+        $sanitizeFlags += " -fno-omit-frame-pointer"            # https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html
```

Comment:
Shouldn't we only enable these flags for debug or for specific pipelines? If sanitizer is enabled during compilation always on these platforms, those hooks get added to the resulting binary, which means we'd ship a binary with sanitizer enabled, making it much slower. We only want sanitizer during specific testing. We might need to make the decision on whether or not to run sanitizer based on which pipeline this is running on.

Reply:
Good point. What do you think about enabling sanitizers in Debug version only?

The Sanitizers documentation announces the slow-down of a factor of 2 - 3 per major sanitizer. If multiple major sanitizers are on (e.g. [`-fsanitize=undefined`](https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html), [`-fsanitize=address`](https://clang.llvm.org/docs/AddressSanitizer.html), ...), then I expect the total slow-down of a factor of 8 - 10. I'm not sure how critical that is at the current stage of QIR RT development.

After I ship the sanitizers, all our tests and samples will be running clean and the sanitizers will only prevent us from the changes that break the clean tests/samples run.
But the users (like PNNL) may be running our QIR RT in the ways we never tested. Their runs may trigger new sanitizers errors/warnings, thus helping us disclose new problem fragments. E.g. recently there was an issue of growing memory use in one of the solvers/algorithms. If the sanitizers were enabled, then in case of a memory leak the [leak sanitizer](https://clang.llvm.org/docs/LeakSanitizer.html) could report something actionalble.

To summarize, I fell that we are interested in getting our users run the _sanitized_ QIR RT. But on the other hand we don't want them to run frustratingly slow. If we enable the sanitizers in the debug version only then they will be able to choose.

Reply:
I think these two sanitizers enabled by default on debug builds makes sense for now. As has come up in discussion, we'll need to figure out what it looks like to run sanitizers on the CI pipeline, as right now CI always builds only the release packages that eventually get published to nuget and runs the tests on those. Furthermore, the Q# packages only build on Windows, such that any tests that are part of that build (such as all of the ones under src/Qir/Tests) are not run on Linux and won't include sanitizers. Taking this together with the fact that not all sanitizers can be enabled together, and I think it makes sense for sanitizer validation to become a specific e2e build validation step (akin to the "Valide end-to-end Linux" step) that tries running tests with different sanitizers enabled in parallel.

As far as customer scenarios go, my two cents is that we should never ship/publish a sanitizer enabled runtime. It's not common nor easy to publish two versions of the same package for the same platform (we'd have to do so under a different name), and even then only whatever sanitizer we pick as default would be enabled. I think it makes sense to compile a debug version with the relevant sanitizer enabled as needed for investigating specific issues rather than try to have a setup that catches everything all the time. There, we may need to compile the release version with sanitizers enabled specifically because debug builds can often change timing or relationship between code due to differing optimizations. Overall, my view is that sanitizers is something we can and should have a dedicated step for in the CI for validation, but otherwise are best used as an on-demand tool rather than an on-by-default tool.

Reply:
Two more issues have been added internally to track the items:
* Ensure that we ship a runtime version built in release without sanitizers enabled
* Ensure all sanitizers and additional validation runs during CI builds